app/gui/src/controller/searcher/input.rs

```diff
@@ -0,0 +1,418 @@
+///
+/// Both information are deduced from the _edited name_. The edited name is a concrete
+/// identifier which we deduce the user is currently editing - we assume this is an identifier where
+/// the text cursor is positioned (including the "end" of the identifier, like `foo|`).
+#[allow(missing_docs)]
+#[derive(Clone, Debug, Default)]
+pub struct Input {
+    /// The input in the AST form.
+    pub ast:             InputAst,
+    /// The current cursor position in the input.
+    pub cursor_position: text::Byte,
+    /// The edited part of the input: the name being edited and it's context. See [`EditedAst`] for
+    /// details.
+    pub edited_ast:      EditedAst,
+}
+
+impl Input {
+    /// Create the structure from the code already parsed to AST.
+    pub fn new(ast: ast::BlockLine<Ast>, cursor_position: text::Byte) -> Self {
+        // We primarily check the character on the left of the cursor (`cursor_position_before`),
+        // to properly handle the situation when the cursor is at end of the edited name.
+        // But if there is no identifier on the left then we check the right side.
+        let cursor_position_before =
+            (cursor_position > text::Byte(0)).then(|| cursor_position - text::ByteDiff(1));
+        let ast_stack_on_left =
+            cursor_position_before.map(|cp| AstAtPositionFinder::find(&ast.elem, cp));
+        let edited_ast_on_left = ast_stack_on_left.map_or_default(EditedAst::new);
+        let edited_ast = if edited_ast_on_left.edited_name.is_none() {
+            let ast_stack_on_right = AstAtPositionFinder::find(&ast.elem, cursor_position);
+            let edited_ast_on_right = EditedAst::new(ast_stack_on_right);
+            if edited_ast_on_right.edited_name.is_some() {
+                edited_ast_on_right
+            } else {
+                edited_ast_on_left
+            }
+        } else {
+            edited_ast_on_left
+        };
+        let ast = InputAst::Line(ast);
+        Self { ast, cursor_position, edited_ast }
+    }
+
+    /// Create the structure parsing the string.
+    pub fn parse(parser: &Parser, expression: impl Str, cursor_position: text::Byte) -> Self {
+        match parser.parse_line(expression.as_ref()) {
+            Ok(ast) => Self::new(ast, cursor_position),
+            Err(_) => Self {
+                ast: InputAst::Invalid(expression.into()),
+                cursor_position,
+                edited_ast: default(),
+            },
+        }
+    }
+
+    /// Return the part of the input used as filtering pattern.
+    pub fn pattern(&self) -> &str {
+        if let Some(edited) = &self.edited_ast.edited_name {
+            let cursor_position_in_name = self.cursor_position - edited.range.start;
+            let name = ast::identifier::name(&edited.ast);
+            name.map_or("", |n| {
+                &n[text::Range::new(text::Byte(0), text::Byte(0) + cursor_position_in_name)]
```

Comment:
This slice can panic if `edited_ast` and `cursor_position` are inconsistent. That *shouldn't* happen, but the invariant relied on is not trivial--slicing with `get` here would reduce the consequences of potential future bugs.